src/collapse/sample.rs

```diff
@@ -0,0 +1,226 @@
+                return Some(false);
+            }
+
+            if line.starts_with(START_LINE) {
+                found_start = true;
+                continue;
+            } else if line.starts_with(END_LINE) {
+                return Some(found_start);
+            }
+        }
+        None
+    }
+}
+
+impl From<Options> for Folder {
+    fn from(opt: Options) -> Self {
+        Folder {
+            opt,
+            ..Default::default()
+        }
+    }
+}
+
+impl Folder {
+    fn line_parts<'a>(&self, line: &'a str) -> Option<(&'a str, &'a str, &'a str)> {
+        let mut line = line.trim_start().splitn(2, ' ');
+        let time = line.next()?.trim_end();
+        let line = line.next()?;
+        let mut line = line.splitn(2, "(in ");
+        let func = line.next()?.trim_end();
+
+        let mut module = "";
+        if !self.opt.no_modules {
+            // Modules are shown with "(in libfoo.dylib)" or "(in AppKit)".
+            // We've arleady split on "(in " above.
+            if let Some(line) = line.next() {
+                if let Some(close) = line.find(')') {
+                    module = &line[..close];
+                }
+
+                // Remove ".dylib", since it adds no value.
+                if module.ends_with(".dylib") {
+                    module = &module[..module.len() - 6]
+                }
+            }
+        }
+
+        Some((time, func, module))
+    }
+
+    fn is_indent_char(c: char) -> bool {
+        c == ' ' || c == '+' || c == '|' || c == ':' || c == '!'
+    }
+
+    // Handle call graph lines of the form:
+    //
+    // 5130 Thread_8749954
+    //    + 5130 start_wqthread  (in libsystem_pthread.dylib) ...
+    //    +   4282 _pthread_wqthread  (in libsystem_pthread.dylib) ...
+    //    +   ! 4282 __doworkq_kernreturn  (in libsystem_kernel.dylib) ...
+    //    +   848 _pthread_wqthread  (in libsystem_pthread.dylib) ...
+    //    +     848 __doworkq_kernreturn  (in libsystem_kernel.dylib) ...
+    fn on_line<W: Write>(&mut self, line: &str, writer: &mut W) -> io::Result<()> {
+        if let Some(indent_chars) = line[4..].find(|c| !Self::is_indent_char(c)) {
+            // Each indent is two characters
+            if indent_chars % 2 != 0 {
+                error!("Odd number of indentation characters for line:\n{}", line);
+            }
+
+            let prev_depth = self.stack.len();
+            let depth = indent_chars / 2 + 1;
+
+            if depth <= prev_depth {
+                self.write_stack(writer)?;
+                for _ in 0..=prev_depth - depth {
+                    self.stack.pop();
+                }
+            } else if depth > prev_depth + 1 {
+                error!("Skipped indentation level at line:\n{}", line);
+            }
+
+            if let Some((samples, func, module)) = self.line_parts(&line[4 + indent_chars..]) {
+                if let Ok(samples) = samples.parse::<usize>() {
+                    self.current_samples = samples;
```

Comment:
This was intentional based on my understanding of how `sample` works. The way I see it, for each sample, whatever function it was in when sampled will get a leaf entry in the call graph. The children counts should always add up to the count of the parent. The function of a parent node could also show up as a leaf node if it was also sampled when it was at the top of the call stack. In the example below, `std::sys::unix::fs::File::open_c` is both a parent node and a leaf node.
```
    +                   !     618 grep_searcher::searcher::Searcher::search_path::h4867309d69f27419  (in rg) + 127  [0x103749bdf]
    +                   !     : 618 std::fs::OpenOptions::_open::h1cd1c180d181e2f5  (in rg) + 33  [0x103a1e481]
    +                   !     :   618 std::sys::unix::fs::File::open::hb90e1c1c787080f0  (in rg) + 198  [0x103a14ad6]
    +                   !     :     617 std::sys::unix::fs::File::open_c::h3660b29577d73852  (in rg) + 338  [0x103a14c72]
    +                   !     :     | 617 std::sys::unix::cvt_r::h0e253dc028f09451  (in rg) + 53  [0x103a1ac55]
    +                   !     :     |   617 __open  (in libsystem_kernel.dylib) + 10  [0x7fff739051ee]
    +                   !     :     1 std::sys::unix::fs::File::open_c::h3660b29577d73852  (in rg) + 372  [0x103a14c94]
```
Am I not understanding this correctly?

Reply:
Oh, that might be completely correct, I just wanted to point it out as I don't know the format used by `sample`. It might be good to verify this (perhaps from the man page) and then document it in the code above.